python/lsst/verify/compatibility/metricsControllerTask.py

```diff
@@ -0,0 +1,420 @@
+    `lsst.verify.compatibility.MetricTask`).
+    """
+
+    # TODO: remove this method in DM-16535
+    def _makeTimingTask(self, target, metric):
+        """Create a `~lsst.verify.compatibility.MetricTask` that can time
+        a specific method.
+
+        Parameters
+        ----------
+        target : `str`
+            the method to time, optionally prefixed by one or more tasks in
+            the format of `lsst.pipe.base.Task.getFullMetadata()`. All
+            matching methods will be counted by the task.
+        metric : `str`
+            the fully qualified name of the metric to contain the
+            timing information
+
+        Returns
+        -------
+        task : `TimingMetricTask`
+            a task that times ``target`` and stores the information in
+            ``metric``. It assumes ``target`` is called as part of
+            `lsst.ap.pipe.ApPipeTask`.
+        """
+        config = TimingMetricTask.ConfigClass()
+        config.metadataDataset = "apPipe_metadata"
+        config.target = target
+        config.metric = metric
+        return TimingMetricTask(config=config)
+
+    def __init__(self, config=None, **kwargs):
+        super().__init__(config=config, **kwargs)
+
+        # TODO: generalize in DM-16535
+        self.measurers = [
+            self._makeTimingTask("apPipe.runDataRef", "ap_pipe.ApPipeTime"),
+            self._makeTimingTask("apPipe:ccdProcessor.runDataRef",
+                                 "pipe_tasks.ProcessCcdTime"),
+            self._makeTimingTask("apPipe:ccdProcessor:isr.runDataRef",
+                                 "ip_isr.IsrTime"),
+            self._makeTimingTask("apPipe:ccdProcessor:charImage.runDataRef",
+                                 "pipe_tasks.CharacterizeImageTime"),
+            self._makeTimingTask("apPipe:ccdProcessor:calibrate.runDataRef",
+                                 "pipe_tasks.CalibrateTime"),
+            self._makeTimingTask("apPipe:differencer.runDataRef",
+                                 "pipe_tasks.ImageDifferenceTime"),
+            self._makeTimingTask("apPipe:differencer:register.run",
+                                 "pipe_tasks.RegisterImageTime"),
+            self._makeTimingTask("apPipe:differencer:measurement.run",
+                                 "ip_diffim.DipoleFitTime"),
+            self._makeTimingTask("apPipe:associator.run",
+                                 "ap_association.AssociationTime"),
+        ]
+
+    @staticmethod
+    def _getInstrument(dataref):
+        """Extract the instrument name from a Butler repo.
+
+        Returns
+        -------
+        The canonical name of the instrument, in all uppercase form.
+        """
+        camera = dataref.get('camera')
+        instrument = camera.getName()
+        return instrument.upper()
+
+    def _computeSingleMeasurement(self, job, metricTask, dataref):
+        """Call a single metric task on a single dataref.
```

Comment:
I'm somewhat concerned that this breaks the expectation in Gen3 that all the IO is done prior to execution.  I suspect that in most cases for verify, the execution is almost entirely this IO operation anyway, but I'm not completely sure this will work cleanly.

Reply:
Everything in the `compatibility` package is specific to Gen 2. This is especially true for `MetricsControllerTask`, which is basically just a stand-in for the dataset-handling capabilities that will be provided by `Pipeline`, `QuantumGraph`, etc.. See DMTN-098 for more details, including a sketch of how migration might work.